File: src/enemies/PapillosaurKing.cpp
/*
 * Copyright (C) 2009 Christopho, Solarus - http://www.solarus-engine.org
 *
 * Solarus is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solarus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include "enemies/PapillosaurKing.h"
#include "entities/MapEntities.h"
#include "movements/RandomWalkMovement.h"
#include "Game.h"
#include "Map.h"
#include "Sprite.h"
#include "SpriteAnimationSet.h"
#include "lowlevel/Random.h"
#include "lowlevel/System.h"
#include "lowlevel/Sound.h"

/**
 * @brief Constructor.
 * @param params name and position of the enemy to create
 */
PapillosaurKing::PapillosaurKing(const ConstructionParameters &params):
  Enemy(params), nb_eggs_to_create(0) {

}

/**
 * @brief Destructor.
 */
PapillosaurKing::~PapillosaurKing() {

}

/**
 * @brief Initializes the enemy.
 */
void PapillosaurKing::initialize() {

  // attack/defense features
  set_damage(2, 0);
  set_life(6);

  // sprite
  create_sprite("enemies/boss_papillosaurking", true);
  set_size(176, 96);
  set_origin(88, 64);
  set_collision_modes(COLLISION_SPRITE);

  // reactions to attacks
  set_no_attack_consequences();
  set_attack_consequence(ATTACK_SWORD, -1);    // protected against sword
  set_attack_consequence(ATTACK_EXPLOSION, 1); // sensible to explosions
}

/**
 * @brief Restarts the normal movement of this enemy.
 *
 * This function is called when the enemy needs to restart its movement
 * because something happened (for example the enemy has just been created,
 * or it was just hurt).
 */
void PapillosaurKing::restart() {
  Enemy::restart();

  // state
  next_egg_phase_date = System::now() + 2500;
  next_egg_date = 0;
  get_sprite()->set_current_animation("walking");

  // movement
  if (movement == NULL) {
    set_movement(new RandomWalkMovement(5));
  }
}

/**
 * @brief Returns whether a destructible item is currently considered as an obstacle for this entity.
 * @param destructible_item a destructible item
 * @return true if the destructible item is currently an obstacle for this entity
 */
bool PapillosaurKing::is_destructible_item_obstacle(DestructibleItem *destructible_item) {
  // as the papillosaur is a big flying boss, it can fly over destructible items
  return false;
}

/**
 * @brief Suspends or resumes the entity.
 * @param suspended true to suspend the entity, false to resume it
 */
void PapillosaurKing::set_suspended(bool suspended) {

  Enemy::set_suspended(suspended);

  if (!suspended) {
    uint32_t diff = System::now() - when_suspended;
    next_egg_phase_date += diff;
    next_egg_date += diff;
  }
}

/**
 * @brief Updates this entity.
 */
void PapillosaurKing::update() {

  Enemy::update();

  if (suspended) {
    return;
  }

  // create minillosaur eggs
  uint32_t now = System::now();
  if (is_in_normal_state() && now >= next_egg_phase_date - 500) {

    // count the number of sons
    std::list<MapEntity*> *sons = map->get_entities()->get_entities_with_prefix(ENEMY, get_name() + "_minillosaur");
    int nb_sons = sons->size();
    if (nb_sons >= 5 && nb_eggs_to_create == 0) {
      // delay the egg phase if there are already too much sons
      next_egg_phase_date += 5000;
    }
    else if (get_movement() != NULL) {
      // stop the movement
      clear_movement();
    }
    else if (now >= next_egg_phase_date) {
      // time to start an egg phase

      if (get_sprite()->get_current_animation() != "preparing_egg") {
	// after a delay, start the fast animation and play a sound
	Sound::play("boss_charge");
	get_sprite()->set_current_animation("preparing_egg");
	next_egg_date = now + 1500;
	nb_eggs_to_create = (get_life() < 3) ? 3 : 1;
      }
      else if (now >= next_egg_date) {

	MapEntity *egg = create(game, MINILLOSAUR, RANK_NORMAL, -1, get_name() + "_minillosaur",
	    get_layer(), get_x(), get_y() + 16, 0, NULL);
	map->get_entities()->add_entity(egg);
	Sound::play("boss_fireball");

	if (--nb_eggs_to_create > 0) {
	  next_egg_date = now + 500;
	}
	else {
	  next_egg_phase_date = now + 5000 + (1 + Random::get_number(3) * 1000);
	  get_sprite()->set_current_animation("walking");
	  set_movement(new RandomWalkMovement(5));
	}
      }
    }
  }
}

/**
 * @brief This function is called when the enemy has just been hurt.
 * @param source the source of the attack
 * @param attack the attack that was just successful
 * @param life_points the number of life points lost by this enemy
 */
void PapillosaurKing::just_hurt(MapEntity *source, EnemyAttack attack, int life_points) {

  if (get_life() <= 0) {
    // the papillosaur is dying: remove the minillosaur eggs
    std::list<MapEntity*> *sons = map->get_entities()->get_entities_with_prefix(ENEMY, get_name() + "_minillosaur");
    for (std::list<MapEntity*>::iterator it = sons->begin(); it != sons->end(); it++) {
      (*it)->remove_from_map();
    }
  }
}

File: quests/zsdx/data/enemies/papillosaur_king.lua
local enemy = ...

-- A big butterfly boss from Newlink.

local nb_eggs_to_create = 0
local nb_eggs_created = 0
local boss_starting_life = 6
local boss_movement_starting_speed = 50  -- Starting speed in pixels per second, it will gain 5 per life point lost.
local boss_movement_speed = boss_movement_starting_speed
local timers = {}

function enemy:on_created()

  self:set_life(boss_starting_life)
  self:set_damage(2)
  self:create_sprite("enemies/papillosaur_king")
  self:set_size(176, 96)
  self:set_origin(88, 64)
  self:set_invincible()
  self:set_attack_consequence("explosion", 1)
  self:set_attack_consequence("sword", "protected")
  self:set_obstacle_behavior("flying")
end

function enemy:on_restarted()

  for _, t in ipairs(timers) do t:stop() end
  timers[#timers + 1] = sol.timer.start(self, 2000, function() self:egg_phase_soon() end)
  self:go()
end

function enemy:on_hurt(attack, life_lost)

  for _, t in ipairs(timers) do t:stop() end
  local life = self:get_life()
  if life <= 0 then
    -- I am dying: remove the minillosaur eggs.
    local sons_prefix = self:get_name() .. "_minillosaur"
    self:get_map():remove_entities(sons_prefix)
  else
    boss_movement_speed = boss_movement_starting_speed
      + (boss_starting_life - life) * 5
  end
end

function enemy:go()
  local m
  if self:get_life() > 1 then
    m = sol.movement.create("random_path")
    m:set_speed(boss_movement_speed)
  else
    -- The enemy is now desperate and angry against our hero.
    m = sol.movement.create("target")
    m:set_speed(boss_movement_speed)
  end
  m:start(self)
end

function enemy:egg_phase_soon()

  local sons_prefix = self:get_name() .. "_minillosaur"
  local nb_sons = self:get_map():get_entities_count(sons_prefix)
  if nb_sons >= 5 then
    -- Delay the egg phase if there are already too much sons.
    timers[#timers + 1] = sol.timer.start(self, 5000, function() self:egg_phase_soon() end)
  else
    self:stop_movement()
    timers[#timers + 1] = sol.timer.start(self, 500, function() self:egg_phase() end)
  end
end

function enemy:egg_phase()

  local sprite = self:get_sprite()
  sprite:set_animation("preparing_egg")
  sol.audio.play_sound("boss_charge")
  timers[#timers + 1] = sol.timer.start(self, 1500, function() self:throw_egg() end)

  -- The more the boss is hurt, the more it will throw eggs...
  nb_eggs_to_create = boss_starting_life - self:get_life() + 1
end

function enemy:throw_egg()

  -- Create the egg.
  nb_eggs_created = nb_eggs_created + 1
  local egg_name = self:get_name() .. "_minillosaur_" .. nb_eggs_created
  local egg = self:create_enemy(egg_name, "minillosaur_egg_thrown", 0, 16)
  egg:set_no_treasure()
  sol.audio.play_sound("boss_fireball")

  -- See what to do next.
  nb_eggs_to_create = nb_eggs_to_create - 1
  if nb_eggs_to_create > 0 then
    -- Throw another egg in 0.5 second.
    timers[#timers + 1] = sol.timer.start(self, 500, function() self:throw_egg() end)
  else
    -- Finish the egg phase.
    local sprite = self:get_sprite()
    sprite:set_animation("walking")
    -- Don't throw eggs when desperate!
    if self:get_life() > 1 then
      -- Schedule the next one in a few seconds.
      local delay = 3500 + (math.random(3) * 1000)
      timers[#timers + 1] = sol.timer.start(self, duration, function() self:egg_phase_soon() end)
    end
    self:go()
  end
end

